test/testfloat/testfloat.cpp

```diff
@@ -0,0 +1,441 @@
+
+        {"f32_to_f64", {32, Type::f64, Type::f32}},
+        {"f64_to_f32", {33, Type::f32, Type::f64}},
+
+        {"f32_ceil", {34, Type::f32, Type::f32}},
+        {"f32_floor", {35, Type::f32, Type::f32}},
+        {"f32_trunc", {36, Type::f32, Type::f32}},
+        {"f32_nearest", {37, Type::f32, Type::f32}},
+        {"f64_ceil", {38, Type::f64, Type::f64}},
+        {"f64_floor", {39, Type::f64, Type::f64}},
+        {"f64_trunc", {40, Type::f64, Type::f64}},
+        {"f64_nearest", {41, Type::f64, Type::f64}},
+    };
+
+    if (const auto it = map.find(name); it != map.end())
+        return it->second;
+    throw std::invalid_argument{"unknown <function>: " + std::string{name}};
+}
+
+constexpr fizzy::Value make_value(Type type, uint64_t bits) noexcept
+{
+    switch (type)
+    {
+    case Type::i32:
+        return fizzy::Value{static_cast<uint32_t>(bits)};
+    case Type::i64:
+        return fizzy::Value{bits};
+    case Type::f32:
+        return fizzy::Value{FP{static_cast<uint32_t>(bits)}.value};
+    case Type::f64:
+        return fizzy::Value{FP{bits}.value};
+    }
+    __builtin_unreachable();
+}
+
+class TypedValue
+{
+    Type m_type;
+    fizzy::Value m_value;
+
+public:
+    constexpr TypedValue(Type type, fizzy::Value value) noexcept : m_type{type}, m_value{value} {}
+
+    constexpr TypedValue(Type type, uint64_t bits) noexcept
+      : m_type{type}, m_value{make_value(type, bits)}
+    {}
+
+    bool eq(uint64_t expected_bits, bool ignore_nans) const
+    {
+        switch (m_type)
+        {
+        case Type::i32:
+        case Type::i64:
+            return m_value.i64 == expected_bits;
+        case Type::f32:
+        {
+            const FP value{m_value.f32};
+            const FP expected{static_cast<uint32_t>(expected_bits)};
+            if (ignore_nans && std::isnan(expected.value))
+            {
+                if (expected.is_canonical_nan())
+                    return value.is_canonical_nan();
+                if (expected.is_arithmetic_nan())
+                    return value.is_arithmetic_nan();
+                throw std::invalid_argument{"invalid input: unexpected signaling NaN"};
+            }
+            return value == expected;
+        }
+
+        case Type::f64:
+        {
+            const FP value{m_value.f64};
+            const FP expected{expected_bits};
+            if (ignore_nans && std::isnan(expected.value))
+            {
+                if (expected.is_canonical_nan())
+                    return value.is_canonical_nan();
+                if (expected.is_arithmetic_nan())
+                    return value.is_arithmetic_nan();
+                throw std::invalid_argument{"invalid input: unexpected signaling NaN"};
+            }
+            return value == expected;
+        }
+        }
+        __builtin_unreachable();
+    }
+
+    friend std::ostream& operator<<(std::ostream& os, const TypedValue& x)
+    {
+        std::ios os_state{nullptr};
```

Comment:
Here we you did that using `flags()`, I think it would be enough here, too
https://github.com/wasmx/fizzy/blob/5e6b32c30e8a754ad1479feb66043b2daa99f1fd/test/utils/asserts.cpp#L15-L20

Reply:
The "flags" will not restore the `setfill`...